src/frontend/src/utils/i18n.ts

```diff
@@ -0,0 +1,62 @@
+/** Internationalization (i18n) and copy support for Internet Identity.
+ *
+ * This module provides helper for loading multi-language copy definitions and updating the lit templates
+ * depending on the user-selected language dynamically. */
+
+import { TemplateResult, html } from "lit-html";
+import { asyncReplace } from "lit-html/directives/async-replace.js";
+import { Chan } from "./utils";
+
+// The type of raw copy, i.e. { lang: { some_key_title: "Welcome!" } }
+type StringCopy<Keys extends string, Lang extends string> = {
+  [key in Lang]: { [key in Keys]: string };
+};
+
+// The dynamic templates, i.e. { some_key_title: <language-dependent template> }
+type DynamicCopy<Keys extends string> = { [key in Keys]: TemplateResult };
+
+/// A class that converts a string copy definition into dynamic lit templates dependent on the
+// currently selected language.
+export class I18n<Lang extends string> {
+  // The 'Chan' used to propagate language updates
+  private chan: Chan<Lang>;
+
+  // The currently selected language
+  private lang: Lang;
+
+  constructor(def: Lang) {
+    this.chan = new Chan(def);
+    this.lang = def;
+  }
+
+  i18n<Keys extends string>(copy: StringCopy<Keys, Lang>): DynamicCopy<Keys> {
+    // Use the default language as the definition for expected keys.
+    // Missing keys in other languages are a type error.
+    type DefaultCopy = StringCopy<Keys, Lang>[Lang];
+    const defCopy: DefaultCopy = copy[this.lang];
+    const keys: [Keys] = Object.keys(defCopy) as [Keys];
```

Comment:
really not sure at all, can `defCopy` be `undefined` here? 

if the `lang` is initialized with a unknown language, couldn't it be the case? if yes, `Ojbect.keys` will throw an error

Reply:
So Typescript would prevent this, because (from the types) `copy` _must_ have a key `this.lang`. Now, that's the theory....

Reply:
Nice then